File: 2096.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
 
        def is_finded(node: TreeNode, value: int, path: List[str]) -> bool:
            if node.val == value:
                return True
            if node.right and is_finded(node.right, value, path):
                path += "R"
            elif node.left and is_finded(node.left, value, path):
                path += "L"
            return path

        s, d = [], []
        is_finded(root, startValue, s)
        is_finded(root, destValue, d)

        while len(s) and len(d) and s[-1] == d[-1]:
            s.pop()
            d.pop()
        return "".join("U" * len(s)) + "".join(reversed(d))

File: Пукаев/2096.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string getDirections(TreeNode* root, int startValue, int destValue) {
        string pathToStart, pathToDest;
        DFS(root, startValue, pathToStart);
        DFS(root, destValue, pathToDest);

        while((pathToStart.size() || pathToDest.size()) && pathToStart.back() == pathToDest.back()){
            pathToStart.pop_back();
            pathToDest.pop_back();
        }

        string result = string(pathToStart.size(), 'U') + pathToDest;
        return result;   
    }


    bool DFS(TreeNode* root, int target, string path){
        if(root->val == target){
            return true;
        }
        if(root->left && DFS(root->left, target, path)) {
            path += 'L';
        }
        else if(root->right && DFS(root->right, target, path)){
            path += 'R';
        } 
        return path.size() > 0;        
    }
};